onnxruntime/contrib_ops/cpu/tokenizer.cc

```diff
@@ -447,6 +493,133 @@ Status Tokenizer::SeparatorTokenize(OpKernelContext* ctx,
+  auto curr_input = input_data;
+  auto const last = input_data + N * C;
+
+  // We do not constraint the search to match
+  // on the beginning or end of the string
+  const RE2::Anchor anchor = RE2::UNANCHORED;
+
+  while (curr_input != last) {
+    const auto& s = *curr_input;
+    tokens.emplace_back();
+    auto& row = tokens.back();
+
+    StringPiece text(s);
+    const auto end_pos = s.length();
+    size_t start_pos = 0;
+    StringPiece submatch;
+
+    bool match = true;
+    while (match) {
+      match = regex_->Match(text, start_pos, end_pos, anchor, &submatch, 1);
+      if (match) {
+        // Record  pos/len
+        assert(submatch.data() != nullptr);
+        size_t match_pos = submatch.data() - s.data();
+        assert(match_pos >= start_pos);
+        auto token_len = match_pos - start_pos;
+        if (token_len > 0) {
+          row.emplace_back(start_pos, token_len);
+        }
+        // Update starting position
+        // Guard against empty string match
+        auto match_len = submatch.length();
+        if (match_len > 0) {
+          start_pos = match_pos + match_len;
+        } else {
+          start_pos = match_pos + 1;
+        }
+      } else {
+        // record trailing token
+        auto trailing_len = end_pos - start_pos;
+        if (trailing_len > 0) {
+          row.emplace_back(start_pos, trailing_len);
+        }
+      }
+    }
+    size_t tokens_num = row.size();
+    if (mark_) {
+      tokens_num += 2;  // Start/end markers as separate tokens
+    }
+    max_tokens = std::max(max_tokens, tokens_num);
+    ++curr_input;
+  }
+  // Check for empty output
+  std::vector<int64_t> output_dims(input_dims);
+  // Check if we have no output due to either empty input
+  // everything is a separator
+  if ((max_tokens - mark_ * 2) == 0) {
+    output_dims.push_back(0);
+    TensorShape output_shape(output_dims);
+    ctx->Output(0, output_shape);
+    return Status::OK();
+  }
+
+  output_dims.push_back(max_tokens);
+  TensorShape output_shape(output_dims);
+
+  auto output_tensor = ctx->Output(0, output_shape);
+  auto const output_data = output_tensor->template MutableData<std::string>();
+
+#ifdef _DEBUG
+  const size_t max_output_index = N * C * max_tokens;
+#endif
+  curr_input = input_data;
+  size_t output_index = 0;
+  for (const auto& row : tokens) {
+    assert(curr_input != last);
+#ifdef _DEBUG
+    size_t c_idx = output_index;
+#endif
+    if (mark_) {
+      (output_data + output_index)->assign(&start_text, 1);
+      ++output_index;
+    }
+    // Output tokens for this row
+    const char* data = curr_input->data();
+    for (const auto& token : row) {
+#ifdef _DEBUG
+      auto s_len = curr_input->length();
+      assert(token.second > 0);
+      assert(token.first < s_len);
+      assert(token.first + token.second <= s_len);
+#endif
+      (output_data + output_index)->assign(data + token.first, token.second);
+      ++output_index;
```

Comment:
I am not sure if it works with multi-byte words in UTF-8.

Reply:
It works